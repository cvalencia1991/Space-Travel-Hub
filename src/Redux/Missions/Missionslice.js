import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const initialState = [];
const GET_MISSIONS = 'GET_MISSIONS';

export const getMissions = createAsyncThunk(
  GET_MISSIONS,
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/missions');
    const result = await response.json();
    return result;
  },
);

const MissionsSlice = createSlice({
  name: 'Missions',
  initialState,
  reducers: {
    Missionstatus(state, action) {
      const newState = state.map((mission) => {
        if (mission.id !== action.payload) { return mission; }
        return { ...mission, reserved: true };
      });
      return newState;
    },
    LeaveMission(state, action) {
      const newState = state.map((mission) => {
        if (mission.id !== action.payload) { return mission; }
        return { ...mission, reserved: false };
      });
      return newState;
    },
  },
  extraReducers: {
    [getMissions.fulfilled]: (state, action) => action.payload.map((element) => (
      {
        id: element.mission_id,
        mission_name: element.mission_name,
        description: element.description,
      }
    )),
  },
});

export const { Missionstatus, LeaveMission } = MissionsSlice.actions;
export default MissionsSlice.reducer;
