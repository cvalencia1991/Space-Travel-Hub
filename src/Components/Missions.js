/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from 'react';
import { Button } from 'react-bootstrap';
import { useSelector, useDispatch } from 'react-redux';
import { getMissions, MissionLeave } from '../Redux/Missions/Missionslice';

let didInit = false;

export default function Missions() {
  const dispatch = useDispatch();
  useEffect(() => {
    if (!didInit) {
      didInit = true;
      dispatch(getMissions());
    }
  }, []);
  const missions = useSelector((state) => state.missions);

  const joinMission = (id) => {
    dispatch(joinMission(id));
  };

  // eslint-disable-next-line no-unused-vars
  const LeaveMission = (id) => {
    dispatch(MissionLeave(id));
  };

  return (
    <table className="table table-bordered table-striped m-5 w-auto p-3">
      <tbody>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th className="stylesjoin"> </th>
        </tr>
        {missions.map((mission) => (
          <tr key={mission.id}>
            <th>{mission.mission_name}</th>
            <th>{mission.description}</th>
            <th>status</th>
            <th><Button onClick={() => joinMission(mission.id)}>Join Mission</Button></th>
          </tr>
        ))}
      </tbody>
    </table>
  );
}
