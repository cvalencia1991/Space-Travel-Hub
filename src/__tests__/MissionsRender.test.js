import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../Components/Missions';
import store from '../Redux/configureStore';

it('renders correctly', () => {
  const missions = renderer
    .create(<Provider store={store}><Missions /></Provider>)
    .toJSON();
  expect(missions).toMatchSnapshot();
});
