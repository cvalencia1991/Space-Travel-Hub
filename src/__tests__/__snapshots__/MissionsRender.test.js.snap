// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="table table-bordered table-striped m-5 w-auto p-3"
>
  <tbody>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th
        className="stylesjoin"
      >
         
      </th>
    </tr>
  </tbody>
</table>
`;
